source/geod24/concurrency.d

```diff
@@ -934,76 +1007,616 @@ protected:
+
+            Unbuffered mode:
+
+                If a writer is already blocked on the Channel, reader copies the
+                value to `output` and wakes up the writer by yielding
+
+                If no writer is ready in the wait queue, reader appends itself
+                to read wait queue and blocks
+
+                If channel is closed, it returns immediatly with a failure
+
+            Buffered mode:
+
+                If there are existing messages in the buffer, reader pops one off
+                the buffer and returns immediatly with success, regardless of the
+                Channel being closed or not
+
+                If there are no messages in the buffer it behaves exactly like the
+                unbuffered operation
+
+            Param:
+                output = Reference to output variable
+            Returns:
+                Success/Failure - Only fails when channel is closed and there are
+                no existing messages to be read.
+
+        ***********************************************************************/
+
+        bool read(ref T output) nothrow
+        {
+            bool success = tryRead(output);
+
+            if (!success && !this.closed)
+            {
+                ChannelQueueEntry q_ent = enqueueEntry(this.readq, &output);
+
+                q_ent.sem.wait();
+                return q_ent.success;
+            }
+
+            return success;
+        }
+
+        /***********************************************************************
+
+            Try to read a message from Channel without blocking
+
+            tryRead locks the Channel and reads a message if it is possible
+            to do so without blocking. If the read would block, it returns without
+            unlocking the Channel to allow caller to `enqueueEntry` without competing
+            for the lock again.
+
+            If the tryRead is being executed in a select call, it tries to consume the
+            `caller_sel` with the given `caller_sel_id`. It only proceeds if it can
+            successfully consume the `caller_sel`
+
+            Param:
+                output = Field to write the message to
+                caller_sel = SelectState instace of the select call being executed
+                caller_sel_id = id of the select call being executed
+
+            Returns:
+                Success/Failure
+
+        ***********************************************************************/
+
+        private bool tryRead(ref T output, SelectState caller_sel = null, int caller_sel_id = 0) nothrow
+        {
+            this.lock.lock_nothrow();
+
+            ChannelQueueEntry write_ent = dequeueEntry(this.writeq, caller_sel, caller_sel_id);
+
+            if (this.buffer.length > 0)
+            {
+                // if dequeueEntry fails, we will try to consume caller_sel again.
+                if (!caller_sel || write_ent || caller_sel.tryConsume(caller_sel_id))
```

Comment:
Won't this segfault if `caller_sel` is null and `write_ent` is true?

Reply:
If `caller_sel` is null, the latter 2 expressions won't get evaluated.

Reply:
Oh right, I misread it! Personally I would make this a bit clearer:

```suggestion
                if (write_ent || (caller_sel && caller_sel.tryConsume(caller_sel_id))
```

I think that would have the same semantics?

Reply:
We still want to execute the body of `if` when `caller_sel` is null. I think this would fail in that case.

Reply:
Ah yep. Carry on. :)